File: 03 Rule Learners/03_association_rules.py

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

#-------------------------------------------------------------------------------------------------#
#---------------------------------------EDA and Cleaning------------------------------------------#
#-------------------------------------------------------------------------------------------------#

raw_d = pd.read_excel("Online Retail.xlsx", sheet_name="Online Retail")
print(raw_d.head())

def naCol(df, print_result=True):
    """
    Checks for null or missing values in all columns of a Pandas dataframe

    Arguments:
    df: A Pandas dataframe
    print_result: indicates whether or not the output should be printed to the console

    Returns:
    dict: (key:value) = (column_name:number_missing_values)    
    """
    y = dict.fromkeys(df.columns)
    for idx, key in enumerate(y.keys()):
        if df.dtypes[list(y.keys())[idx]] == 'object':
            y[key] = pd.isnull(df[list(y.keys())[idx]]).sum() + (df[list(y.keys())[idx]]=='').sum()
        else:
            y[key] = pd.isnull(df[list(y.keys())[idx]]).sum()
    if print_result:
        print("Number of nulls by column:")
        for k, v in y.items():
            print(k, v)
    return y

naCol(raw_d)
# Documentation says InvoiceNo starting with "C" is cancellation
print(len(raw_d[raw_d['InvoiceNo'].astype('str').apply(lambda x: x[:1])=='C']))
print(set(raw_d['InvoiceNo'].astype('str').apply(lambda x: x[:1])))  # What does "A" mean?
print(raw_d[raw_d['InvoiceNo'].astype('str').apply(lambda x: x[:1])=='A'][:6])

# Are there any transactions with unusual quantities?
print(raw_d.Quantity.value_counts())
print(raw_d[raw_d['Quantity']<0][:6])     # Possibly returns
print(raw_d[raw_d['Quantity']>1000][:6])  # Cheap items in bulk

# Are there any unusual unit prices?
sns.distplot(raw_d['UnitPrice'])
plt.show()
print(raw_d[raw_d['UnitPrice']<0][:6])     # Adjustments
print(raw_d[raw_d['UnitPrice']>1000][:6])  # Postage and Amazon Fees
# Consider removing StockCodes in ['AmazonFee', 'M' , 'Post', 'Dot', B']

# Can a customer have multiple invoices?
print(raw_d[['InvoiceNo', 'CustomerID']].groupby('CustomerID').agg('count'))

# Does an invoice contain multiple items?
items_by_invoice = raw_d[['StockCode', 'InvoiceNo']].groupby('InvoiceNo').agg('count').reset_index()
print(items_by_invoice)
# Are there invoices with only 1 item?  These can be removed.
invoices_to_remove = items_by_invoice[items_by_invoice['StockCode']<2]['InvoiceNo'].tolist()

# Remove data that could produce meaningless or useless rules
clean_d = raw_d[(~raw_d['InvoiceNo'].str.slice(0,1).isin(['C', 'A'])) \
                & (raw_d['Quantity']>0) \
                & (raw_d['UnitPrice']>0) \
                & (~raw_d['StockCode'].isin(['AMAZONFEE', 'M', 'POST', 'DOT', 'B'])) \
                & (~raw_d['InvoiceNo'].isin(invoices_to_remove))] \
                [['InvoiceNo', 'StockCode', 'Description']]

print(clean_d.info())
clean_d['InvoiceNo'] = clean_d['InvoiceNo'].astype('int')
clean_d['Description'] = clean_d['Description'].str.strip()

#clean_d.to_csv('online_retail_clean.csv', index=False)

#-------------------------------------------------------------------------------------------------#
#------------------------------------Apriori Association Rules------------------------------------#
#-------------------------------------------------------------------------------------------------#

trxn_d = clean_d[['InvoiceNo', 'Description']].groupby('InvoiceNo')['Description'].apply(list).reset_index()
te = TransactionEncoder()
te_ary = te.fit(trxn_d['Description']).transform(trxn_d['Description'])
d = pd.DataFrame(te_ary, columns=te.columns_)

# Sense checks
print(len(set(clean_d['Description'])))  #should match nbr rows
print(len(set(clean_d['InvoiceNo'])))    #should match nbr columns
print(clean_d.info())
print(d.info())
print(d.sum()[d.sum()<1])
print(d.sum()[d.sum()==np.inf])
print(d.sum(axis=1)[d.sum(axis=1)<1])
print(d.sum(axis=1)[d.sum(axis=1)==np.inf])

frequent_itemsets = apriori(d, min_support=0.03,
                            use_colnames=False,
                            max_len=None)
print(frequent_itemsets.head())
a_rules = association_rules(frequent_itemsets,
                            metric="confidence",
                            min_threshold=0.7)
print(a_rules.head())
#a_rules.to_csv('online_retail_learned_rules.csv', index=False)

File: 03 Rule Learners/03_association_rules.R
# https://archive.ics.uci.edu/ml/datasets/online+retail

setwd("dsilt-ml-code/03 Rule Learners")

#install.packages("openxlsx", dependencies = TRUE)
library(openxlsx)
library(Matrix)
library(arules)
library(arulesViz)

set.seed(14)

#-------------------------------------------------------------------------------------------------#
#---------------------------------------EDA and Cleaning------------------------------------------#
#-------------------------------------------------------------------------------------------------#

raw_d <- read.xlsx('Online Retail.xlsx', sheet='Online Retail')
head(raw_d)

naCol <- function(x){
  y <- sapply(x, function(y) sum(length(which(is.na(y) | is.nan(y) | is.infinite(y) | y=='NA' | y=='NaN' | y=='Inf' | y==''))))
  y <- data.frame('feature'=names(y), 'count.nas'=y)
  row.names(y) <- c()
  y
}

naCol(raw_d)
# Documentation says InvoiceNo starting with "C" is cancellation
nrow(raw_d[substring(as.character(raw_d$InvoiceNo), 1, 1)=='C',])
unique(substring(as.character(raw_d$InvoiceNo), 1, 1))  # What does "A" mean?
raw_d[substring(as.character(raw_d$InvoiceNo), 1, 1)=="A", 1:6]

# Are there any transactions with unusual quantities?
table(raw_d$Quantity)
raw_d[raw_d$Quantity<0, 1:6]    # Possibly returns
raw_d[raw_d$Quantity>1000, 1:6] # Cheap items in bulk

# Are there any unusual unit prices?
hist(raw_d$UnitPrice)
raw_d[raw_d$UnitPrice<0, 1:6]    # Adjustments
raw_d[raw_d$UnitPrice>1000, 1:6] # Postage and Amazon Fees
# Consider removing StockCodes in ['AmazonFee', 'M' , 'Post', 'Dot', B']

# Can a customer have multiple invoices?
aggregate(InvoiceNo ~ CustomerID, data=raw_d, FUN=length)

# Does an invoice contain multiple items?
items_by_invoice <- aggregate(StockCode ~ InvoiceNo, data=raw_d, FUN=length)
items_by_invoice
# Are there invoices with only 1 item?  These can be removed.
invoices_to_remove <- items_by_invoice[items_by_invoice$StockCode<2, 'InvoiceNo']

# Remove data that could produce meaningless or useless rules
clean_d <- raw_d[!(substring(as.character(raw_d$InvoiceNo), 1, 1) %in% c('C', 'A'))
                 & (raw_d$Quantity>0)
                 & (raw_d$UnitPrice>0)
                 & !(raw_d$StockCode %in% c('AMAZONFEE', 'M', 'POST', 'DOT', 'B'))
                 & !(raw_d$InvoiceNo %in% invoices_to_remove),
                 c('InvoiceNo', 'StockCode', 'Description')]
clean_d <- clean_d[complete.cases(clean_d),]

sapply(clean_d, class)
clean_d$InvoiceNo <- as.integer(clean_d$InvoiceNo)
clean_d$Description <- trimws(clean_d$Description)

write.csv(clean_d, 'online_retail_clean.csv', row.names=F)

#-------------------------------------------------------------------------------------------------#
#------------------------------------Apriori Association Rules------------------------------------#
#-------------------------------------------------------------------------------------------------#

# Cast to sparse matrix
trxns <- as(split(clean_d[, "Description"], 
                  clean_d[, "InvoiceNo"]), 
            "transactions")
dm <- as(trxns, "ngCMatrix")  # ngCMatrix is more efficient than dgCMatrix

# Sparse matrix sense checks
length(unique(clean_d$Description)) #should match nbr rows
length(unique(clean_d$InvoiceNo))   #should match nbr columns
str(trxns)
str(dm)
sum(colSums(dm, na.rm=F, dims = 1, sparseResult=F)<1)
sum(colSums(dm, na.rm=F, dims = 1, sparseResult=F)==Inf)
sum(rowSums(dm, na.rm=F, dims = 1, sparseResult=F)<1)
sum(rowSums(dm, na.rm=F, dims = 1, sparseResult=F)==Inf)

# Inspect part of the matrix and look for any obvious patterns
arules::image(dm[1:2000, 1:2000])

#Default hyperparameters: minimum sup=0.5 and conf=0.8 were not satisfactory
dmrules <- apriori(dm, parameter=list(sup=0.03, conf=0.1, target="rules", minlen=2))
inspect(dmrules)

dmrules <- apriori(dm, parameter=list(sup=0.02, conf=0.7, target="rules", minlen=2))
inspect(dmrules)

dmrules <- apriori(dm, parameter=list(sup=0.01, conf=0.8, target="rules", minlen=2))
inspect(head(sort(dmrules, by="lift"), 25))

dmrules_df <- as(sort(dmrules, by="lift"), "data.frame")
write.csv(dmrules_df, 'online_retail_learned_rules.csv', row.names=F)

# Visualizing rules
plot(dmrules, measure=c('support', 'lift'), shading='confidence', interactive=T)
plot(head(dmrules, 10), method='grouped')
plot(dmrules, method='graph')
plot(dmrules, method='graph', control=list(type='itemsets'))
plot(dmrules, method='paracoord')
#plot(dmrules, method='paracoord', control=list(reorder=T))  #Takes a long time to reorder
# Documentation here: https://rdrr.io/rforge/arulesViz/f/inst/doc/arulesViz.pdf
